### PR TITLE
Update assert to handle `object[]` element type

### LINKED ISSUE #105186: `JIT/IL_Conformance/Convert/TestConvertFromIntegral/TestConvertFromIntegral` failing in CI with "Assert failure: `arrayBaseRef->GetArrayElementType() == ELEMENT_TYPE_OBJECT`"
Example pipeline run (gcstress in this case): https://dev.azure.com/dnceng-public/public/_build/results?buildId=748823&view=results
Example console log: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-105181-merge-0ee628e940694b27bb/IL_Conformance.0.1/1/console.238ae2ed.log?helixlogtype=result

```scala
16:32:00.456 Running test: JIT/IL_Conformance/Convert/TestConvertFromIntegral/TestConvertFromIntegral.dll

Assert failure(PID 22 [0x00000016], Thread: 22 [0x0016]): arrayBaseRef->GetArrayElementType() == ELEMENT_TYPE_OBJECT
    File: /__w/1/s/src/coreclr/vm/comutilnative.cpp:438
    Image: /root/helix/work/correlation/corerun

...

Stack trace:
----------------------------------
Thread Id: 0x16
      Child SP               IP Call Site
 0xffffdc241e80 0xffffa8028c0c libc.so.6!wait4
 0xffffdc241ec0 0xffffa7e42bec libcoreclr.so!PROCCreateCrashDump(std::vector<char const*, std::allocator<char const*> >&, char*, int, bool) at /__w/1/s/src/coreclr/pal/src/thread/process.cpp:0:0
 0xffffdc241f30 0xffffa7e43ce8 libcoreclr.so!PROCCreateCrashDumpIfEnabled at /__w/1/s/src/coreclr/pal/src/thread/process.cpp:2526:9
 0xffffdc241fb0 0xffffa7e40bf4 libcoreclr.so!PROCAbort at /__w/1/s/src/coreclr/pal/src/thread/process.cpp:2559:5
 0xffffdc241fd0 0xffffa7e40a70 libcoreclr.so!CorUnix::TerminateCurrentProcessNoExit(int) at /__w/1/s/src/coreclr/pal/src/thread/process.cpp:3240:0
 0xffffdc243270 0xffff67747044 0xffff67747044
 0xffffdc243350 0xffff6774711c System.Private.CoreLib.dll!System.Exception.get_TargetSite()
 0xffffdc243370 0xffff67747c44 System.Private.CoreLib.dll!System.Exception.get_Source()
 0xffffdc243390 0xffff67747168 System.Private.CoreLib.dll!System.Exception.InternalPreserveStackTrace()
 0xffffdc2433b0 0xffffa7c220b8 0xffffa7c220b8
 0xffffdc2433d0 0xffffa79ede98 libcoreclr.so!NativeExceptionHolderBase::~NativeExceptionHolderBase() at ??:0:0
 (inlined by) CallDescrWorkerWithHandler(CallDescrData*, int) at ??:65:5
 0xffffdc243430 0xffffa79ee72c libcoreclr.so!MethodDescCallSite::CallTargetWorker(unsigned long const*, unsigned long*, int) at ??:0:0
 0xffffdc2436a0 0xffffa78bceec libcoreclr.so!ExceptionPreserveStackTrace(OBJECTREF) at ??:0:0
 0xffffdc2437f0 0xffffa7b5b2b4 libcoreclr.so!DispatchManagedException(OBJECTREF, _CONTEXT*, bool) at ??:0:0
 0xffffdc244bf0 0xffffa7b5af7c libcoreclr.so!DispatchManagedException(PAL_SEHException&, bool) at ??:0:0
 0xffffdc245040 0xffffa7ae4c50 libcoreclr.so!RuntimeMethodHandle::InvokeMethod(Object*, void**, SignatureNative*, bool) at ??:0:0
 0xffffdc245540 0xffff679a8bfc System.Private.CoreLib.dll!System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(System.Object, System.Reflection.BindingFlags)
 0xffffdc245580 0xffff679efb7c System.Private.CoreLib.dll!System.Reflection.Emit.DynamicMethod.Invoke(System.Object, System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo)
 0xffffdc2455d0 0xffff6a2c7280 TestConvertFromIntegral.dll!TestCasts.Program.GenerateTest[[System.Int32, System.Private.CoreLib],[System.SByte, System.Private.CoreLib]](Int32, System.Reflection.Emit.OpCode, System.Reflection.Emit.OpCode, Boolean, SByte, Boolean)
 0xffffdc245650 0xffff6a2c6d84 TestConvertFromIntegral.dll!TestCasts.Program.TestConvertFromInt4ToI1()
 0xffffdc245680 0xffff6a2c6a7c TestConvertFromIntegral.dll!TestCasts.Program.TestConvertFromInt4()
 0xffffdc245690 0xffff6a2c696c TestConvertFromIntegral.dll!TestCasts.Program.TestEntryPoint()
 0xffffdc2456a0 0xffff6a2c6620 IL_Conformance.dll!Program.<<Main>$>g__TestExecutor399|0_400(System.IO.StreamWriter, System.IO.StreamWriter, <>c__DisplayClass0_0 ByRef)
 0xffffdc245720 0xffff68fb5640 IL_Conformance.dll!Program.<Main>$(System.String[])
```

@AaronRobinsonMSFT Could this be related to #105103?

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Fixes https://github.com/dotnet/runtime/issues/105186

@jakobbotsch this is very odd. This code path is the same on all platforms, but only seems to be taken with these tests on non-Windows platforms. That is the reason it didn't fail on Windows. Any ideas why exceptions are being thrown on non-Windows, but not on Windows?

### CODE REVIEW

File: src/coreclr/vm/comutilnative.cpp [hunk 1]
```diff
@@ -433,9 +433,10 @@ extern "C" void QCALLTYPE ExceptionNative_GetMethodFromStackTrace(QCall::ObjectH
         // The stacktrace can be either sbyte[] or Object[]. In the latter case,
         // the first entry is the actual stack trace sbyte[], the rest are pointers
         // to the method info objects. We only care about the first entry here.
-        if (arrayBaseRef->GetArrayElementType() != ELEMENT_TYPE_I1)
+        CorElementType elemType = arrayBaseRef->GetArrayElementType();
+        if (elemType != ELEMENT_TYPE_I1)
         {
-            _ASSERTE(arrayBaseRef->GetArrayElementType() == ELEMENT_TYPE_OBJECT);
+            _ASSERTE(elemType == ELEMENT_TYPE_OBJECT || elemType == ELEMENT_TYPE_CLASS);
```
Comment (jkotas) [inline src/coreclr/vm/comutilnative.cpp:439]:
```suggestion
            _ASSERTE(elemType == ELEMENT_TYPE_CLASS); // object[]
```
ELEMENT_TYPE_OBJECT should only ever show up in signatures. We should never see it here.

**Review by jkotas — APPROVED** [review]

**Comment by jakobbotsch** [comment]
> Any ideas why exceptions are being thrown on non-Windows, but not on Windows?

The JIT is generating directy call to `CORINFO_HELP_OVERFLOW` for these tests on both Windows and Linux, so seems like there is something different there. It is bound to `JIT_Overflow` that does `COMPlusThrow(kOverflowException);`, but apparently that does not end up in this code on Windows.
Maybe @janvorli has better idea what is different here.

BTW, while stepping through this on Windows I ended up in this weird code:
https://github.com/dotnet/runtime/blob/98b165db27a3c15b9c0df208d1acca573b3dd15e/src/coreclr/vm/excep.cpp#L7354-L7357

**Comment by janvorli** [comment]
> BTW, while stepping through this on Windows I ended up in this weird code:

I've noticed that recently too. This is a remainder of logging of a piece of stack into stresslog, see this in coreclr 3.1:
https://github.com/dotnet/coreclr/blob/v3.1.32/src/vm/excep.cpp#L8139-L8147.
The body of the loop was removed , but blame on github somehow isn't able to show me the change that did that.

**Comment by jkotas** [comment]
> his is a remainder of logging of a piece of stack into stresslog, see this in coreclr 3.1:
https://github.com/dotnet/coreclr/blob/v3.1.32/src/vm/excep.cpp#L8139-L8147.
The body of the loop was removed , but blame on github somehow isn't able to show me the change that did that.

This was deleted in 
https://github.com/dotnet/runtime/pull/81927/files#diff-47cbfa47fb88845e5e76fd4df0a84e9da494508ffe80da116511fc2d0becd9baL7614

Deleting the rest #105346

**Comment by AaronRobinsonMSFT** [comment]
> Maybe @janvorli has better idea what is different here.

I tried to debug this and it is mostly related to the SEH simulation on non-Windows. Particularly usage of `NativeExceptionHolderBase`.

**Pull request merged**